WellLog.Lib/Models/DLIS/ComponentDescriptor.cs

```diff
@@ -0,0 +1,50 @@
+﻿using WellLog.Lib.Helpers;
+
+namespace WellLog.Lib.Models.DLIS
+{
+    public class ComponentDescriptor
+    {
+        private const byte ABSENT_ATTRIBUTE_ROLE = 0b_0000_0000;
+        private const byte ATTRIBUTE_ROLE = 0b_0010_0000;
+        private const byte INVARIANT_ATTRIBUTE_ROLE = 0b_0100_0000;
+        private const byte OBJECT_ROLE = 0b_0110_0000;
+        private const byte REDUNDANT_SET_ROLE = 0b_1010_0000;
+        private const byte REPLACEMENT_SET_ROLE = 0b_1100_0000;
+        private const byte SET_ROLE = 0b_1110_0000;
+
+        private const byte SET_TYPE_MASK = 0b_0001_0000;
+        private const byte SET_NAME_MASK = 0b_0000_1000;
+
+        private const byte OBJ_NAME_MASK = 0b_0001_0000;
+
+        private const byte ATTR_LABEL_MASK = 0b_0001_0000;
+        private const byte ATTR_COUNT_MASK = 0b_0001_0000;
+        private const byte ATTR_REPRESENTATION_CODE_MASK = 0b_0001_0000;
+        private const byte ATTR_UNITS_MASK = 0b_0001_0000;
+        private const byte ATTR_VALUE_MASK = 0b_0001_0000;
+
+        private readonly byte _descriptor;
+
+        public bool IsAbsentAttribute => _descriptor.IsComponentRole(ABSENT_ATTRIBUTE_ROLE);
+        public bool IsAttribute => _descriptor.IsComponentRole(ATTRIBUTE_ROLE);
+        public bool IsInvariantAttribute => _descriptor.IsComponentRole(INVARIANT_ATTRIBUTE_ROLE);
+        public bool IsObject => _descriptor.IsComponentRole(OBJECT_ROLE);
+        public bool IsRedundantSet => _descriptor.IsComponentRole(REDUNDANT_SET_ROLE);
+        public bool IsReplacementSet => _descriptor.IsComponentRole(REPLACEMENT_SET_ROLE);
+        public bool IsSet => _descriptor.IsComponentRole(SET_ROLE);
+
+        public bool SetHasType => (IsSet || IsRedundantSet || IsReplacementSet) ? _descriptor.GetBit(SET_TYPE_MASK) : false;
+        public bool SetHasName => (IsSet || IsRedundantSet || IsReplacementSet) ? _descriptor.GetBit(SET_NAME_MASK) : false;
+        public bool ObjectHasName => IsObject ? _descriptor.GetBit(OBJ_NAME_MASK) : false;
+        public bool AttributeHasLabel => (IsAttribute || IsInvariantAttribute) ? _descriptor.GetBit(ATTR_LABEL_MASK) : false;
+        public bool AttributeHasCount => (IsAttribute || IsInvariantAttribute) ? _descriptor.GetBit(ATTR_COUNT_MASK) : false;
+        public bool AttributeHasRepresentationCode => (IsAttribute || IsInvariantAttribute) ? _descriptor.GetBit(ATTR_REPRESENTATION_CODE_MASK) : false;
+        public bool AttributeHasUnits => (IsAttribute || IsInvariantAttribute) ? _descriptor.GetBit(ATTR_UNITS_MASK) : false;
+        public bool AttributeHasValue => (IsAttribute || IsInvariantAttribute) ? _descriptor.GetBit(ATTR_VALUE_MASK) : false;
```

Comment:
nit: These lines are "clever" code and difficult to read because they're in a single line. "Clever code" or code that's difficult to read will just be skipped over. Which is where nasty bugs lie because it doesn't really get read. I'd recommend moving these to functions.

Reply:
i would debate the notion that a simple inline conditional is either clever or difficult to read.

@FyLMartin or @TomMarks , thoughts?

Reply:
I think it's fairly straightforward.  Lambda expressions/arrow functions are the norm now, especially with the rise of JS as a real language.

Reply:
I think the ternary operator can be simplified to:

```cs
=> (IsAttribute || IsInvariantAttribute) && _descriptor.GetBit(ATTR_UNITS_MASK);
```

Reply:
you just increased the confusion in the name of "simplification".  making your solution more "clever" than mine